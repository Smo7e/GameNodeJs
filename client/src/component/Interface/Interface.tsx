import React, { useContext, useEffect, useRef, useState } from "react";

import { MediatorContext, StoreContext } from "../../App";
import { TGamer, TMob, TMobs } from "../../modules/Server/types";
import { act } from "react-dom/test-utils";

import Chat from "./component/Chat/Chat";
import ParametersGame from "./component/ParametersGame/ParametersGame";
import TaskSelection from "./component/TaskSelection/TaskSelection";

import "./Interface.css";
import { VARIABLE } from "../../modules/Store/Store";
import CheatMenu from "./component/CheatMenu/CheatMenu";
import WinOrLose from "./component/WinOrLose/WinOrLose";
import Music from "../Parameters/Music";

const Interface: React.FC = () => {
    const mediator = useContext(MediatorContext);
    const store = useContext(StoreContext);

    const [mobs, setMobs] = useState<TMobs | null>(null);
    const timeRef = useRef<HTMLDivElement>(null);
    const [gamers, setGamers] = useState<TGamer[] | null>(null);
    const [timer, setTimer] = useState({ seconds: 0, minutes: 0 });
    const parametersGameRef = useRef<HTMLDivElement>(null);
    const [showParametersGame, setShowParametersGame] = useState(false);
    const cheatMenu = true;
    const handleSettingsClick = () => {
        setShowParametersGame(true);
    };

    const handleOutsideClick = (event: MouseEvent) => {
        if (parametersGameRef.current && !parametersGameRef.current.contains(event.target as Node)) {
            setShowParametersGame(false);
        }
    };
    useEffect(() => {
        document.addEventListener("mousedown", handleOutsideClick);
        const updateTime = () => {
            setTimer((prevTimer) => {
                let seconds = prevTimer.seconds + 1;
                let minutes = prevTimer.minutes;

                if (seconds === 60) {
                    minutes++;
                    seconds = 0;
                }

                return { seconds, minutes };
            });
        };
        let timeInterval = setInterval(updateTime, 1000);
        return () => {
            document.removeEventListener("mousedown", handleOutsideClick);
            clearInterval(timeInterval);
        };
    }, []);

    useEffect(() => {
        const { GET_GAMERS } = mediator.getEventTypes();
        const { GET_MOBS } = mediator.getEventTypes();

        const getGamersHandler = (data: TGamer[]) => {
            setGamers(data);
        };
        const getMobsHandler = (data: TMobs) => {
            setMobs(data);
        };

        mediator.subscribe(GET_GAMERS, getGamersHandler);
        mediator.subscribe(GET_MOBS, getMobsHandler);

        return () => {
            mediator.unsubscribe(GET_GAMERS, getGamersHandler);
            mediator.unsubscribe(GET_MOBS, getMobsHandler);
        };
    });

    // вызов окошка задания

    return (
        <div className="Interface-container">
            {/* <Music /> */}
            <div className="back-arrow-interface"></div>
            <div onClick={handleSettingsClick} className="settings-arrow-interface"></div>
            {showParametersGame && (
                <div ref={parametersGameRef}>
                    <ParametersGame />
                </div>
            )}

            <div className="player-info">
                {gamers && gamers[0].person_id === 0 ? (
                    <div className="player-icon-interfaceSportik"></div>
                ) : gamers && gamers[0].person_id === 2 ? (
                    <div className="player-icon-interfaceHumanitarian"></div>
                ) : gamers && gamers[0].person_id === 1 ? (
                    <div className="player-icon-interfaceTheechguy"></div>
                ) : (
                    <></>
                )}

                <div className="I">
                    {gamers ? (
                        gamers.map((elem: TGamer, index: number) => (
                            <div
                                key={index}
                                className="HP"
                                style={{
                                    background: `linear-gradient(to right, red ${elem.hp}%, #898888 ${elem.hp}%)`,
                                }}
                            >
                                {elem.name} : {elem.hp}/100HP
                            </div>
                        ))
                    ) : (
                        <></>
                    )}
                </div>

                <div className="player-level-interface">1</div>
            </div>

            <div className="map-info">
                <div className="miniMap-interface">Мини карта</div>

                <div className="game-info-interface">
                    <div className="coin-interface">0</div>
                    <div className="coin-icon-interface"></div>
                    <div ref={timeRef} className="time">
                        {timer.minutes < 10 ? "0" + timer.minutes : timer.minutes} :{" "}
                        {timer.seconds < 10 ? "0" + timer.seconds : timer.seconds}
                    </div>
                </div>
            </div>
            <></>
            {store.get(VARIABLE.TRIGGERTRUSOV) ? (
                <>
                    <div className="BossXP">{mobs ? <div>TrusovHP: {mobs["trusov"].hp}</div> : <></>}</div>
                    <TaskSelection mobName={"trusov"} />
                </>
            ) : (
                <></>
            )}
            {store.get(VARIABLE.TRIGGERRUSANOVA) ? (
                <>
                    <div className="BossXP">{mobs ? <div>RusanovaHP: {mobs["rusanova"].hp}</div> : <></>}</div>
                    <TaskSelection mobName={"rusanova"} />
                </>
            ) : (
                <></>
            )}
            {store.get(VARIABLE.TRIGGERGOLOVIZIN) ? (
                <>
                    <div className="BossXP">{mobs ? <div>GolovizinHP: {mobs["golovizin"].hp}</div> : <></>}</div>
                    <TaskSelection mobName={"golovizin"} />
                </>
            ) : (
                <></>
            )}

            <Chat />
            <WinOrLose />

            {cheatMenu ? <CheatMenu /> : <></>}
        </div>
    );
};

export default Interface;
