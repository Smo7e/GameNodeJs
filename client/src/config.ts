// UdSU
// const DOMAIN = 'http://studfront';
// const PORT = 81;

// dev
const DOMAIN = "http://localhost";
const PORT = 3001;

// prod
//...

export const HOST = PORT ? `${DOMAIN}:${PORT}` : `${DOMAIN}`;

export const MEDIATOR = {
    EVENTS: {
        SERVER_ERROR: "SERVER_ERROR",
        GET_MESSAGES: "GET_MESSAGES",
        GET_SCENE: "GET_SCENE",
        GET_USER: "GET_USER",
        GET_GAMERS: "GET_GAMERS",
        GET_FRIENDS: "GET_FRIENDS",
        GET_INVITES: "GET_INVITES",
        LOGIN: "LOGIN",
        SIGNUP: "SIGNUP",
        LOGOUT: "LOGOUT",
        GET_MOBS: "GET_MOBS",
        GET_ERROR: "GET_ERROR",
        PLAY_MUSIC: "PLAY_MUSIC", 
        STOP_MUSIC: "STOP_MUSIC", 
        SET_MUSIC_VOLUME: "SET_MUSIC_VOLUME", 
        UPDATE_ARR_BULLET_TRAJECTORY: "UPDATE_ARR_BULLET_TRAJECTORY",
    },
    TRIGGERS: {},
};
