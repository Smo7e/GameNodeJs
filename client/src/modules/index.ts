import { TError } from "./Server/types";
import Server from "./Server/Server";
import Mediator from "./Mediator/Mediator";
import Store from "./Store/Store";

export type { TError };
export { Server, Mediator, Store };
