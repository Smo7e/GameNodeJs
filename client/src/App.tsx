import React, { useState } from "react";
import { HOST, MEDIATOR } from "./config";
import { Server, Mediator, Store, TError } from "./modules";
import SignUp from "./component/SignUp/SignUp";
import Game from "./modules/Game/Game";
import Menu from "./component/Menu/Menu";
import Login from "./component/Login/Login";
import Heroes from "./component/Heroes/Heroes";
import Interface from "./component/Interface/Interface";
import Lobby from "./component/Lobby/Lobby";
import ErrorMessage from "./modules/ErrorMessage/ErrorMessage";
import Parameters from "./component/Parameters/Parameters";
import Music from "./component/Parameters/Music";
import "./App.css";

export const ServerContext = React.createContext<Server>(null!);
export const MediatorContext = React.createContext<Mediator>(null!);
export const StoreContext = React.createContext<Store>(null!);

export enum EPAGES {
    SIGNUP,
    LOGIN,
    GAME,
    MENU,
    HEROES,
    LOBBY,
    PARAMETERS,
}

const MainApp = () => {
    const [epages, setEpages] = useState<EPAGES>(EPAGES.LOGIN);
    return (
        <>
            {epages === EPAGES.LOGIN ? (
                <Login epages={setEpages} />
            ) : epages === EPAGES.SIGNUP ? (
                <SignUp epages={setEpages} />
            ) : epages === EPAGES.MENU ? (
                <Menu epages={setEpages} />
            ) : epages === EPAGES.GAME ? (
                <>
                    <Game />
                    <Interface />
                </>
            ) : epages === EPAGES.HEROES ? (
                <Heroes epages={setEpages} />
            ) : epages === EPAGES.LOBBY ? (
                <Lobby epages={setEpages} />
            ) : epages === EPAGES.PARAMETERS ? (
                <Parameters epages={setEpages} />
            ) : (
                <></>
            )}
            <ErrorMessage />
            <Music />
        </>
    );
};

const App: React.FC = () => {
    const store = new Store();
    const mediator = new Mediator(MEDIATOR);
    const server = new Server(HOST, mediator, store);

    const { SERVER_ERROR } = mediator.getEventTypes();
    mediator.subscribe(SERVER_ERROR, (data: TError) => {
        console.log(data);
    });

    return (
        <StoreContext.Provider value={store}>
            <MediatorContext.Provider value={mediator}>
                <ServerContext.Provider value={server}>
                    <MainApp />
                </ServerContext.Provider>
            </MediatorContext.Provider>
        </StoreContext.Provider>
    );
};

export default App;
