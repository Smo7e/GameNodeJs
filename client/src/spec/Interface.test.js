import { render, screen } from "@testing-library/react";
import Interface from "../component/Interface/Interface";
import { MediatorContext, ServerContext } from "../App";
import { Mediator, Server, Store } from "../modules";
import { HOST, MEDIATOR } from "../config";

const store = new Store();
const mediator = new Mediator(MEDIATOR);
const server = new Server(HOST, mediator, store);
jest.useFakeTimers();

test("Interface snapshot", () => {
    const { asFragment } = render(
        <MediatorContext.Provider value={mediator}>
            <ServerContext.Provider value={server}>
                {/* <Interface /> */}
            </ServerContext.Provider>
        </MediatorContext.Provider>
        
    );

    expect(asFragment()).toMatchSnapshot();

    jest.advanceTimersByTime(1000);
    expect(asFragment()).toMatchSnapshot();

    jest.useRealTimers();
});
