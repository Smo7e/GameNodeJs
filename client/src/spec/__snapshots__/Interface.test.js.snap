// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interface snapshot 1`] = `<DocumentFragment />`;

exports[`Interface snapshot 2`] = `<DocumentFragment />`;
